consul/pool_discovery.go

```diff
@@ -0,0 +1,207 @@
+package consul
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+	"github.com/hashicorp/consul/api"
+	"github.com/hashicorp/nomad-autoscaler/ha"
+	"github.com/hashicorp/nomad-autoscaler/sdk/helper/hash"
+)
+
+// AgentID satisfies the AgentID function on the PoolDiscovery interface.
+func (c *Consul) AgentID() string { return c.agentID }
+
+// MonitorPool satisfies the MonitorPool function on the PoolDiscovery
+// interface.
+func (c *Consul) MonitorPool(ctx context.Context, req ha.DiscoveryRequest) {
+	q := &api.QueryOptions{
+		WaitIndex: 0,
+		WaitTime:  5 * time.Minute,
+	}
+	for {
+
+		// Generate a child context and configure the Consul query options to
+		// utilise this. It allows us to cancel the blocking query as a result
+		// of reloads or shutdowns.
+		consulCtx, consulCancel := context.WithCancel(ctx)
+		q = q.WithContext(consulCtx)
+
+		errCh := make(chan error)
+		resCh := make(chan []*api.CatalogService)
+
+		go func() {
+
+			c.clientLock.RLock()
+			defer c.clientLock.RUnlock()
+
+			// This routine writes to the channels, so therefore is responsible
+			// for closing them on exit.
+			defer close(errCh)
+			defer close(resCh)
+
+			services, qm, err := c.client.Catalog().Service(c.serviceName, autoscalerConsulServiceTag, q)
+
+			// If the context has been cancelled, exit.
+			if consulCtx.Err() != nil {
+				return
+			}
+
+			if err != nil {
+				errCh <- err
+			} else {
+				q.WaitIndex = qm.LastIndex
+				resCh <- services
+			}
+		}()
```

Comment:
Would it be better to have these run just once instead of recreating everything on each loop? So there would be one long-lived goroutine with a `for` loop and another loop surrounding the `select` below